src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -512,33 +523,112 @@ private static ItemGroup getItemGroup(Run<?, ?> build) {
     }
 
 
-    private boolean perform(Run src, Run<?,?> dst, String expandedFilter, String expandedExcludes, FilePath targetDir,
-            FilePath baseTargetDir, Copier copier, PrintStream console)
-            throws IOException, InterruptedException {
-        FilePath srcDir = selector.getSourceDirectory(src, console);
+    private boolean perform(Run src, Run<?,?> dst, String expandedFilter, @CheckForNull String expandedExcludes, FilePath targetDir, PrintStream console) throws IOException, InterruptedException {
+        VirtualFile srcDir = selector.getArtifacts(src, console);
         if (srcDir == null) {
             return isOptional();  // Fail build unless copy is optional
         }
 
-        copier.initialize(src, dst, srcDir, baseTargetDir);
+        Map<String, String> fingerprints;
+        MessageDigest md5;
+        if (isFingerprintArtifacts()) {
+            fingerprints = new HashMap<>();
+            try {
+                md5 = MessageDigest.getInstance("MD5");
+            } catch (NoSuchAlgorithmException x) {
+                throw new AssertionError(x);
+            }
+        } else {
+            fingerprints = null;
+            md5 = null;
+        }
         try {
-            int cnt;
-            if (!isFlatten())
-                cnt = copier.copyAll(srcDir, expandedFilter, expandedExcludes, targetDir, isFingerprintArtifacts());
-            else {
-                targetDir.mkdirs();  // Create target if needed
-                FilePath[] list = srcDir.list(expandedFilter, expandedExcludes, false);
-                for (FilePath file : list)
-                    copier.copyOne(file, new FilePath(targetDir, file.getName()), isFingerprintArtifacts());
-                cnt = list.length;
+            targetDir.mkdirs();  // Create target if needed
+            List<String> list = scan(srcDir, expandedFilter, expandedExcludes);
+            for (String file : list) {
+                copyOne(src, dst, fingerprints, srcDir.child(file), new FilePath(targetDir, isFlatten() ? file.replaceFirst(".+/", "") : file), md5);
```

Comment:
Note for me: snipping ".+/" is equivalent to the current implementation calling `FilePath#getName()` in `FingerprintingCopyMethod`.
(I'm not sure this is a good behavior as reported in [JENKINS-16987](https://issues.jenkins-ci.org/browse/JENKINS-16987))

Reply:
No, this `replaceFirst` call just implements the flatten mode. It has nothing to do with fingerprinting. The behavior you mention is retained in `copyOne` as

```java
fingerprints.put(s.getName(), digest);
//                ^^^^^^^^^^
```

Reply:
Oops, I misplaced the comment!
https://github.com/jenkinsci/copyartifact-plugin/pull/100/files#r174750602